pkg/declcfg/addTocfg.go

```diff
@@ -0,0 +1,104 @@
+package declcfg
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+)
+
+type AddConfigRequest struct {
+	ConfigsDir string
+	OutDir     string
+	Bundles    []string
+	Mode       string
+}
+
+type BundleAdder struct {
+	Logger      *logrus.Entry
+	ImgRegistry image.Registry
+}
+
+func NewBundleAdder(logger *logrus.Entry, imgRegistry image.Registry) BundleAdder {
+	return BundleAdder{
+		Logger:      logger,
+		ImgRegistry: imgRegistry,
+	}
+}
+
+func (b BundleAdder) AddToConfig(request AddConfigRequest) error {
+	b.Logger.Infof("loading configs from directory")
+	decCfg, err := LoadDir(request.ConfigsDir)
+	if err != nil {
+		return fmt.Errorf("Error loading directory:%s", request.ConfigsDir)
+	}
+	simpleRefs := make([]image.Reference, 0)
+	for _, ref := range request.Bundles {
+		simpleRefs = append(simpleRefs, image.SimpleReference(ref))
+	}
+	for _, ref := range simpleRefs {
+		imgRef, unpackDir, cleanup, err := unpackImage(context.TODO(), b.ImgRegistry, ref)
+		if err != nil {
+			return err
+		}
+		defer cleanup()
+
+		img, err := registry.NewImageInput(imgRef, unpackDir)
+		if err != nil {
+			return fmt.Errorf("Error interpreting bundle image:%s", err)
+		}
+		m, err := registry.CreateModelFromBundle(img.Bundle)
+		if err != nil {
+			return fmt.Errorf("Error creating model from bundle:%s", err)
+		}
+		bundleConfig := ConvertFromModel(*m)
+		if !existingPackage(decCfg, img.Bundle.Package) {
+			b.Logger.Infof("adding new package metadata to configs")
+			decCfg.Packages = append(decCfg.Packages, bundleConfig.Packages...)
+		}
+		b.Logger.Infof("adding bundle metadata to configs")
+		decCfg.Bundles = append(decCfg.Bundles, bundleConfig.Bundles...)
+	}
+	os.RemoveAll(request.OutDir) // this is required because WriteDir only writes to an empty directory
+	os.Mkdir(request.OutDir, os.ModePerm)
+	b.Logger.Infof("Writing modified configs to %s", request.OutDir)
+	return WriteDir(*decCfg, request.OutDir)
```

Comment:
This will always rewrite config dirs into our "preferred" structure. There are probably some benefits to that, but I could also see users wanting us to not touch their existing structure.

One potential option would be to search the configs directory for the file containing the `olm.package` for this bundle, and write a new file with just this `olm.bundle` blob.

Reply:
@joelanford this takes us back to our discussion about `WriteFile` vs `writeFile` in the `declCfg` package. I thought the decision was to only expose `WriteFile`, which would translate to: If you use any tool that uses the `declCfg` package, your directory will get rewritten. 

TBH I am in favor of controlling how the directory is laid out, to enforce some hygiene.